// parse/lib/node/CoreManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/** Based on https://github.com/react-native-async-storage/async-storage/blob/main/packages/default-storage-backend/src/types.ts */
/*:: export type WebSocketController = {
  onopen: () => void,
  onmessage: (message: any) => void,
  onclose: (arg?: any) => void,
  onerror: (error: any) => void,
  send: (data: any) => void,
  close: () => void,
};*/
const config /*: Config & { [key: string]: any }*/ = {
  IS_NODE: typeof process !== 'undefined' && !!process.versions && !!process.versions.node && !process.versions.electron,
  REQUEST_ATTEMPT_LIMIT: 5,
  REQUEST_BATCH_SIZE: 20,
  REQUEST_HEADERS: {},
  SERVER_URL: 'https://api.parse.com/1',
  SERVER_AUTH_TYPE: null,
  SERVER_AUTH_TOKEN: null,
  LIVEQUERY_SERVER_URL: null,
  ENCRYPTED_KEY: null,
  VERSION: 'js' + "5.1.0",
  APPLICATION_ID: null,
  JAVASCRIPT_KEY: null,
  MASTER_KEY: null,
  USE_MASTER_KEY: false,
  PERFORM_USER_REWRITE: true,
  FORCE_REVOCABLE_SESSION: false,
  ENCRYPTED_USER: false,
  IDEMPOTENCY: false,
  ALLOW_CUSTOM_OBJECT_ID: false,
  PARSE_ERRORS: []
};
function requireMethods(name /*: string*/, methods /*: Array<string>*/, controller /*: any*/) {
  methods.forEach(func => {
    if (typeof controller[func] !== 'function') {
      throw new Error(`${name} must implement ${func}()`);
    }
  });
}
const CoreManager = {
  get: function (key /*: string*/) /*: any*/{
    if (config.hasOwnProperty(key)) {
      return config[key];
    }
    throw new Error('Configuration key not found: ' + key);
  },
  set: function (key /*: string*/, value /*: any*/) /*: void*/{
    config[key] = value;
  },
  setIfNeeded: function (key /*: string*/, value /*: any*/) /*: any*/{
    if (!config.hasOwnProperty(key)) {
      config[key] = value;
    }
    return config[key];
  },
  /* Specialized Controller Setters/Getters */

  setAnalyticsController(controller /*: AnalyticsController*/) {
    requireMethods('AnalyticsController', ['track'], controller);
    config['AnalyticsController'] = controller;
  },
  getAnalyticsController() /*: AnalyticsController*/{
    return config['AnalyticsController'];
  },
  setCloudController(controller /*: CloudController*/) {
    requireMethods('CloudController', ['run', 'getJobsData', 'startJob'], controller);
    config['CloudController'] = controller;
  },
  getCloudController() /*: CloudController*/{
    return config['CloudController'];
  },
  setConfigController(controller /*: ConfigController*/) {
    requireMethods('ConfigController', ['current', 'get', 'save'], controller);
    config['ConfigController'] = controller;
  },
  getConfigController() /*: ConfigController*/{
    return config['ConfigController'];
  },
  setCryptoController(controller /*: CryptoController*/) {
    requireMethods('CryptoController', ['encrypt', 'decrypt'], controller);
    config['CryptoController'] = controller;
  },
  getCryptoController() /*: CryptoController*/{
    return config['CryptoController'];
  },
  setEventEmitter(eventEmitter /*: any*/) {
    config['EventEmitter'] = eventEmitter;
  },
  getEventEmitter() /*: any*/{
    return config['EventEmitter'];
  },
  setFileController(controller /*: FileController*/) {
    requireMethods('FileController', ['saveFile', 'saveBase64'], controller);
    config['FileController'] = controller;
  },
  setEventuallyQueue(controller /*: EventuallyQueue*/) {
    requireMethods('EventuallyQueue', ['poll', 'save', 'destroy'], controller);
    config['EventuallyQueue'] = controller;
  },
  getEventuallyQueue() /*: EventuallyQueue*/{
    return config['EventuallyQueue'];
  },
  getFileController() /*: FileController*/{
    return config['FileController'];
  },
  setInstallationController(controller /*: InstallationController*/) {
    requireMethods('InstallationController', ['currentInstallationId', 'currentInstallation', 'updateInstallationOnDisk'], controller);
    config['InstallationController'] = controller;
  },
  getInstallationController() /*: InstallationController*/{
    return config['InstallationController'];
  },
  setLiveQuery(liveQuery /*: any*/) {
    config['LiveQuery'] = liveQuery;
  },
  getLiveQuery() /*: any*/{
    return config['LiveQuery'];
  },
  setObjectController(controller /*: ObjectController*/) {
    requireMethods('ObjectController', ['save', 'fetch', 'destroy'], controller);
    config['ObjectController'] = controller;
  },
  getObjectController() /*: ObjectController*/{
    return config['ObjectController'];
  },
  setObjectStateController(controller /*: ObjectStateController*/) {
    requireMethods('ObjectStateController', ['getState', 'initializeState', 'removeState', 'getServerData', 'setServerData', 'getPendingOps', 'setPendingOp', 'pushPendingState', 'popPendingState', 'mergeFirstPendingState', 'getObjectCache', 'estimateAttribute', 'estimateAttributes', 'commitServerChanges', 'enqueueTask', 'clearAllState'], controller);
    config['ObjectStateController'] = controller;
  },
  getObjectStateController() /*: ObjectStateController*/{
    return config['ObjectStateController'];
  },
  setPushController(controller /*: PushController*/) {
    requireMethods('PushController', ['send'], controller);
    config['PushController'] = controller;
  },
  getPushController() /*: PushController*/{
    return config['PushController'];
  },
  setQueryController(controller /*: QueryController*/) {
    requireMethods('QueryController', ['find', 'aggregate'], controller);
    config['QueryController'] = controller;
  },
  getQueryController() /*: QueryController*/{
    return config['QueryController'];
  },
  setRESTController(controller /*: RESTController*/) {
    requireMethods('RESTController', ['request', 'ajax'], controller);
    config['RESTController'] = controller;
  },
  getRESTController() /*: RESTController*/{
    return config['RESTController'];
  },
  setSchemaController(controller /*: SchemaController*/) {
    requireMethods('SchemaController', ['get', 'create', 'update', 'delete', 'send', 'purge'], controller);
    config['SchemaController'] = controller;
  },
  getSchemaController() /*: SchemaController*/{
    return config['SchemaController'];
  },
  setSessionController(controller /*: SessionController*/) {
    requireMethods('SessionController', ['getSession'], controller);
    config['SessionController'] = controller;
  },
  getSessionController() /*: SessionController*/{
    return config['SessionController'];
  },
  setStorageController(controller /*: StorageController*/) {
    if (controller.async) {
      requireMethods('An async StorageController', ['getItemAsync', 'setItemAsync', 'removeItemAsync', 'getAllKeysAsync'], controller);
    } else {
      requireMethods('A synchronous StorageController', ['getItem', 'setItem', 'removeItem', 'getAllKeys'], controller);
    }
    config['StorageController'] = controller;
  },
  setLocalDatastoreController(controller /*: LocalDatastoreController*/) {
    requireMethods('LocalDatastoreController', ['pinWithName', 'fromPinWithName', 'unPinWithName', 'getAllContents', 'clear'], controller);
    config['LocalDatastoreController'] = controller;
  },
  getLocalDatastoreController() /*: LocalDatastoreController*/{
    return config['LocalDatastoreController'];
  },
  setLocalDatastore(store /*: any*/) {
    config['LocalDatastore'] = store;
  },
  getLocalDatastore() {
    return config['LocalDatastore'];
  },
  getStorageController() /*: StorageController*/{
    return config['StorageController'];
  },
  setAsyncStorage(storage /*: AsyncStorageType*/) {
    config['AsyncStorage'] = storage;
  },
  getAsyncStorage() {
    return config['AsyncStorage'];
  },
  setWebSocketController(controller /*: new (url: string | URL, protocols?: string | string[] | undefined) => WebSocketController*/) {
    config['WebSocketController'] = controller;
  },
  getWebSocketController() /*: new (url: string | URL, protocols?: string | string[] | undefined) => WebSocketController*/{
    return config['WebSocketController'];
  },
  setUserController(controller /*: UserController*/) {
    requireMethods('UserController', ['setCurrentUser', 'currentUser', 'currentUserAsync', 'signUp', 'logIn', 'become', 'logOut', 'me', 'requestPasswordReset', 'upgradeToRevocableSession', 'requestEmailVerification', 'verifyPassword', 'linkWith'], controller);
    config['UserController'] = controller;
  },
  getUserController() /*: UserController*/{
    return config['UserController'];
  },
  setLiveQueryController(controller /*: LiveQueryControllerType*/) {
    requireMethods('LiveQueryController', ['setDefaultLiveQueryClient', 'getDefaultLiveQueryClient', '_clearCachedDefaultClient'], controller);
    config['LiveQueryController'] = controller;
  },
  getLiveQueryController() /*: LiveQueryControllerType*/{
    return config['LiveQueryController'];
  },
  setHooksController(controller /*: HooksController*/) {
    requireMethods('HooksController', ['create', 'get', 'update', 'remove'], controller);
    config['HooksController'] = controller;
  },
  getHooksController() /*: HooksController*/{
    return config['HooksController'];
  },
  setParseOp(op /*: typeof ParseOp*/) {
    config['ParseOp'] = op;
  },
  getParseOp() {
    return config['ParseOp'];
  },
  setParseObject(object /*: typeof ParseObject*/) {
    config['ParseObject'] = object;
  },
  getParseObject() /*: ParseObject*/{
    return config['ParseObject'];
  },
  setParseQuery(query /*: typeof ParseQuery*/) {
    config['ParseQuery'] = query;
  },
  getParseQuery() /*: ParseQuery*/{
    return config['ParseQuery'];
  },
  setParseRole(role /*: typeof ParseRole*/) {
    config['ParseRole'] = role;
  },
  getParseRole() /*: ParseRole*/{
    return config['ParseRole'];
  },
  setParseUser(user /*: typeof ParseUser*/) {
    config['ParseUser'] = user;
  },
  getParseUser() /*: ParseUser*/{
    return config['ParseUser'];
  }
};
module.exports = CoreManager;
var _default = CoreManager;
exports.default = _default;
